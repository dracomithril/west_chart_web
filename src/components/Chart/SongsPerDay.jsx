/**
 * Created by Gryzli on 29.03.2017.
 */
import React from 'react';
import PropTypes from 'prop-types';
import FontAwesomeIcon from '@fortawesome/react-fontawesome';
import ExpansionPanel from '@material-ui/core/ExpansionPanel';
import ExpansionPanelDetails from '@material-ui/core/ExpansionPanelDetails';
import ExpansionPanelSummary from '@material-ui/core/ExpansionPanelSummary';
import ExpandMoreIcon from '@material-ui/icons/ExpandMore';
import { faArrowCircleDown, faArrowCircleUp, faExclamationTriangle } from '@fortawesome/fontawesome-free-solid';
import { withStyles } from '@material-ui/core/styles';
import Typography from '@material-ui/core/Typography';
import TextField from '@material-ui/core/TextField';
import './chart.css';
import { errorDaysObjectProps } from '../typeDefinitions';

const styles = theme => ({
  button: {},
  typography: {
    margin: theme.spacing.unit * 2,
  },
  heading: {
    fontSize: theme.typography.pxToRem(15),
    fontWeight: theme.typography.fontWeightRegular,
  },
  root: {
    width: '100%',
  },
  textField: {
    marginLeft: theme.spacing.unit,
    marginRight: theme.spacing.unit,
  },
  error: {
    color: '#a94442',
    backgroundColor: '#f2dede',
    borderColor: '#ebccd1',
  },
  good: {
    color: '#3c763d',
    backgroundColor: '#dff0d8',
    borderColor: '#d6e9c6',
  },
});

const SongsPerDay = props => {
  const { errorDays, songsPerDay, onDaysChange, classes } = props;
  const errorDaysMap = errorDays.map(({ org, color }) => (
    <div key={org}>
      <FontAwesomeIcon icon={color === 'blue' ? faArrowCircleDown : faArrowCircleUp} color={color} />
      <span style={{ color }}>{new Date(org).toDateString()}</span>
    </div>
  ));
  const style = errorDays.length !== 0 ? classes.error : classes.good;

  return (
    <ExpansionPanel className={`${classes.root} ${style}`}>
      <ExpansionPanelSummary expandIcon={<ExpandMoreIcon />}>
        <Typography className={classes.heading}>
          {songsPerDay} songs per day
          {errorDaysMap.length > 0 && <FontAwesomeIcon icon={faExclamationTriangle} color="red" />}
        </Typography>
      </ExpansionPanelSummary>
      <ExpansionPanelDetails>
        <div>
          <TextField
            id="number"
            label="songs per day"
            value={songsPerDay}
            onChange={({ target }) => {
              onDaysChange && onDaysChange(Number(target.value));
            }}
            type="number"
            className={classes.textField}
            InputLabelProps={{
              shrink: true,
            }}
            margin="normal"
          />
          {errorDaysMap}
        </div>
      </ExpansionPanelDetails>
    </ExpansionPanel>
  );
};

SongsPerDay.propTypes = {
  classes: PropTypes.shape({
    button: PropTypes.string,
    typography: PropTypes.string,
  }).isRequired,
  errorDays: PropTypes.arrayOf(errorDaysObjectProps),
  songsPerDay: PropTypes.number,
  onDaysChange: PropTypes.func,
};

export default withStyles(styles)(SongsPerDay);
