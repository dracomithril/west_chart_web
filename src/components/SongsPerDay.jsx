/**
 * Created by Gryzli on 29.03.2017.
 */
import React from 'react';
import PropTypes from 'prop-types';
import { OverlayTrigger, Popover } from 'react-bootstrap';

const DayEntry = ({ org, color }) => (
  <div key={org}>
    <span style={{ color }}>{new Date(org).toDateString()}</span>
  </div>
);
DayEntry.propTypes = {
  org: PropTypes.string,
  color: PropTypes.string,
};

const ErrorDaysIndicator = ({ error_days, less }) =>
  error_days.length > 0 && (
    <OverlayTrigger
      trigger={['hover', 'focus']}
      placement="bottom"
      overlay={
        <Popover id="haveMoreOrLessDays" title={`${less ? 'less' : 'more'} then expected`}>
          {error_days}
        </Popover>
      }
    >
      <i
        className={`fa fa-arrow-circle-down ${less ? 'less_days' : 'more_days'}`}
        aria-hidden="true"
      >
        {error_days.length}
      </i>
    </OverlayTrigger>
  );
ErrorDaysIndicator.propTypes = {
  error_days: PropTypes.array,
  less: PropTypes.bool,
};
const SongsPerDay = ({ error_days, songs_per_day, onDaysChange }) => {
  const err_days_less = error_days.filter(elem => elem.color === 'blue').map(DayEntry);
  const err_days_more = error_days.filter(elem => elem.color === 'red').map(DayEntry);
  return (
    <div
      id="songsPerDay"
      className={error_days.length !== 0 ? 'songsPerDay_err' : 'songsPerDay_good'}
    >
      <div style={{ width: '80%', float: 'left', paddingTop: 10 }}>
        <label
          style={{
            float: 'left',
            paddingTop: 3,
          }}
          htmlFor="songs_per_day"
        >
          {'songs per day'}
          <input
            type="number"
            value={songs_per_day}
            className="num_days"
            style={{ width: 30 }}
            name="songs_per_day"
            id="songs_per_day"
            step={1}
            max={10}
            min={1}
            onChange={({ target }) => {
              onDaysChange(Number(target.value));
            }}
          />
        </label>
      </div>
      <div style={{ width: '20%', float: 'left', paddingTop: 5 }}>
        <ErrorDaysIndicator error_days={err_days_less} less />
        <ErrorDaysIndicator error_days={err_days_more} />
      </div>
    </div>
  );
};
SongsPerDay.propTypes = {
  error_days: PropTypes.array,
  songs_per_day: PropTypes.number,
  onDaysChange: PropTypes.func,
};

export default SongsPerDay;
