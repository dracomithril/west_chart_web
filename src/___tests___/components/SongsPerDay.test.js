/**
 * Created by XKTR67 on 4/19/2017.
 */
import React from 'react';
import PropTypes from 'prop-types';
import SongsPerDay from './../../components/SongsPerDay';
import Enzyme, {shallow} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import {shallowToJson} from 'enzyme-to-json';
import configureMockStore from 'redux-mock-store';

Enzyme.configure({ adapter: new Adapter() });
const initial_state = require('../data/initial_state.json');

const mockStore = configureMockStore([]);
describe('<SongsPerDay/>', () => {
  it('renders without crashing', () => {
    const store = mockStore(initial_state);
    const wrapper = shallow(
      <SongsPerDay error_days={[]}/>, {
        context: { store },
        childContextTypes: { store: PropTypes.object }
      }
    );
    expect(shallowToJson(wrapper)).toMatchSnapshot();

  });
});
