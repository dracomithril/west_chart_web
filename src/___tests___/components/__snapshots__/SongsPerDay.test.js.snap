// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SongsPerDay/> renders without crashing 1`] = `
<div
  className="songsPerDay_good"
  id="songsPerDay"
>
  <div
    style={
      Object {
        "float": "left",
        "paddingTop": 10,
        "width": "80%",
      }
    }
  >
    <label
      htmlFor="songs_per_day"
      style={
        Object {
          "float": "left",
          "paddingTop": 3,
        }
      }
    >
      songs per day
      <input
        className="num_days"
        id="songs_per_day"
        max={10}
        min={1}
        name="songs_per_day"
        onChange={[Function]}
        step={1}
        style={
          Object {
            "width": 30,
          }
        }
        type="number"
      />
    </label>
  </div>
  <div
    style={
      Object {
        "float": "left",
        "paddingTop": 5,
        "width": "20%",
      }
    }
  >
    <ErrorDaysIndicator
      error_days={Array []}
      less={true}
    />
    <ErrorDaysIndicator
      error_days={Array []}
    />
  </div>
</div>
`;
